### ISSUE TITLE
Use Go's builtin map

### ISSUE THREAD

**Issue by tylertreat** [nats-io/gnatsd#164]
I believe `HashMap` was written because Go's pre 1.0 `map` was slow, but benchmarking with Go 1.5.2, shows the builtin `map` is actually faster:

```
go version go1.5.2 darwin/amd64
Benchmark_GoMap___GetSmallKey-8 200000000                6.67 ns/op
Benchmark_HashMap_GetSmallKey-8 100000000               15.2 ns/op
Benchmark_GoMap____GetMedKey-8  200000000                8.15 ns/op
Benchmark_HashMap__GetMedKey-8  100000000               10.8 ns/op
Benchmark_GoMap____GetLrgKey-8  200000000                6.48 ns/op
Benchmark_HashMap__GetLrgKey-8  100000000               24.1 ns/op
Benchmark_GoMap_________Set-8   50000000                35.2 ns/op
Benchmark_HashMap_______Set-8   10000000               113 ns/op
```

The builtin `map` requires `string` keys instead of `[]byte`, so it's possible the `string` conversion makes this difference inconsequential. I replaced the cache in `sublist` and benchmarked the difference:

**sublist with HashMap:**
```
go version go1.5.2 darwin/amd64
Benchmark______________________Insert-8  3000000               517 ns/op
Benchmark____________MatchSingleToken-8 30000000                54.0 ns/op
Benchmark______________MatchTwoTokens-8 30000000                54.1 ns/op
Benchmark____________MatchThreeTokens-8 20000000                57.7 ns/op
Benchmark_____________MatchFourTokens-8 20000000                59.1 ns/op
Benchmark_MatchFourTokensSingleResult-8 20000000                59.2 ns/op
Benchmark_MatchFourTokensMultiResults-8 20000000                57.4 ns/op
Benchmark_______MissOnLastTokenOfFive-8 20000000                62.0 ns/op
```

**sublist with builtin map:**
```
go version go1.5.2 darwin/amd64
Benchmark______________________Insert-8  3000000               513 ns/op
Benchmark____________MatchSingleToken-8 30000000                48.5 ns/op
Benchmark______________MatchTwoTokens-8 30000000                48.9 ns/op
Benchmark____________MatchThreeTokens-8 30000000                51.2 ns/op
Benchmark_____________MatchFourTokens-8 20000000                58.1 ns/op
Benchmark_MatchFourTokensSingleResult-8 20000000                58.1 ns/op
Benchmark_MatchFourTokensMultiResults-8 20000000                58.0 ns/op
Benchmark_______MissOnLastTokenOfFive-8 20000000                60.2 ns/op
```

As you can see, the difference in `sublist` is pretty marginal, but I'm wondering if it's really worth using a custom hashmap anymore.

**Comment by derekcollison** [comment]
It's been faster for awhile! Been my todo list to remove and also address your concurrency concerns for sublist.






On Sat, Dec 12, 2015 at 12:30 PM -0800, "Tyler Treat" <notifications@github.com> wrote:












I believe HashMap was written because Go's pre 1.0 map was slow, but benchmarking with Go 1.5.2, shows the builtin map is actually faster:

go version go1.5.2 darwin/amd64
Benchmark_GoMap___GetSmallKey-8 200000000                6.67 ns/op
Benchmark_HashMap_GetSmallKey-8 100000000               15.2 ns/op
Benchmark_GoMap____GetMedKey-8  200000000                8.15 ns/op
Benchmark_HashMap__GetMedKey-8  100000000               10.8 ns/op
Benchmark_GoMap____GetLrgKey-8  200000000                6.48 ns/op
Benchmark_HashMap__GetLrgKey-8  100000000               24.1 ns/op
Benchmark_GoMap_________Set-8   50000000                35.2 ns/op
Benchmark_HashMap_______Set-8   10000000               113 ns/op




The builtin map requires string keys instead of []byte, so it's possible the string conversion makes this difference inconsequential. I replaced the cache in sublist and benchmarked the difference:



sublist with HashMap:

go version go1.5.2 darwin/amd64
Benchmark______________________Insert-8  3000000               517 ns/op
Benchmark____________MatchSingleToken-8 30000000                54.0 ns/op
Benchmark______________MatchTwoTokens-8 30000000                54.1 ns/op
Benchmark____________MatchThreeTokens-8 20000000                57.7 ns/op
Benchmark_____________MatchFourTokens-8 20000000                59.1 ns/op
Benchmark_MatchFourTokensSingleResult-8 20000000                59.2 ns/op
Benchmark_MatchFourTokensMultiResults-8 20000000                57.4 ns/op
Benchmark_______MissOnLastTokenOfFive-8 20000000                62.0 ns/op




sublist with builtin map:

go version go1.5.2 darwin/amd64
Benchmark______________________Insert-8  3000000               513 ns/op
Benchmark____________MatchSingleToken-8 30000000                48.5 ns/op
Benchmark______________MatchTwoTokens-8 30000000                48.9 ns/op
Benchmark____________MatchThreeTokens-8 30000000                51.2 ns/op
Benchmark_____________MatchFourTokens-8 20000000                58.1 ns/op
Benchmark_MatchFourTokensSingleResult-8 20000000                58.1 ns/op
Benchmark_MatchFourTokensMultiResults-8 20000000                58.0 ns/op
Benchmark_______MissOnLastTokenOfFive-8 20000000                60.2 ns/op




As you can see, the difference in sublist is pretty marginal, but I'm wondering if it's really worth using a custom hashmap anymore.

**Comment by tylertreat** [comment]
Thanks, Derek! I have a branch where I started replacing `HashMap`. Let me know if you think this is worth pursuing: https://github.com/nats-io/gnatsd/compare/master...tylertreat:builtin_map?expand=1

**Comment by derekcollison** [comment]
Will take a look. Want to solve for the high concurrency issue as well.

**Comment by derekcollison** [comment]
Fixed with df02bc0bcf90717889aa9ab9189162fb43d6dc3a.

I have a PR pending that allows an L1 cache with no contention in the client's readLoop as well. Also since the sublist is subscription aware, can optimize queue subscribers, which right now have dismal performance relative to straight pub/sub.
